src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -655,7 +656,7 @@ private string GeneratePropMetadataInitFunc(TypeGenerationSpec typeGenerationSpe
                         : "jsonPropertyName: null";
 
                     string getterNamedArg = memberMetadata.CanUseGetter
-                        ? $"getter: static (obj) => (({declaringTypeCompilableName})obj).{clrPropertyName}"
+                        ? $"getter: static (obj) => (({declaringTypeCompilableName})obj).{clrPropertyName}!"
```

Comment:
Please see https://github.com/dotnet/runtime/issues/52227#issuecomment-896376092 for why this is banged away, rather than changing the [parameter type](https://github.com/dotnet/runtime/blob/db2b87e3ed28fd5833c243e5c567ba34520dc176/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.cs#L42) from `Func<object, T>` to `Func<object, T?>`.

Reply:
I didn't quite follow the explanation.  What goes poorly if you remove the `!`?

Reply:
If I remove the `!`, then I get errors like this:

```
D:\repos\dotnet_runtime_4\src\libraries\System.Text.Json\tests\System.Text.Json.SourceGeneration.Tests\System.Text.Json.SourceGeneration\System.Text.Json.SourceGeneration.JsonSourceGenerator\CollectionTestsContext_Metadata.DictionaryEntry.g.cs(75,45): error CS8603: Possible null reference return. [D:\repos\dotnet_runtime_4\src\libraries\System.Text.Json\tests\System.Text.Json.SourceGeneration.Tests\System.Text.Json.SourceGeneration.Tests.csproj]
```

Here's the generated code responsible for the error:

```
properties[1] = global::System.Text.Json.Serialization.Metadata.JsonMetadataServices.CreatePropertyInfo<global::System.Object>(
  options,
  isProperty: true,
  isPublic: true,
  isVirtual: false,
  declaringType: typeof(global::System.Collections.DictionaryEntry),
  propertyTypeInfo: jsonContext.Object,
  converter: null,
  getter: static (obj) => ((global::System.Collections.DictionaryEntry)obj).Value,
  setter: static (obj, value) => global::System.Runtime.CompilerServices.Unsafe.Unbox<global::System.Collections.DictionaryEntry>(obj).Value = value,
  ignoreCondition: null,
  hasJsonInclude: false,
  numberHandling: default,
  propertyName: "Value",
  jsonPropertyName: null);
```

The issue is the getter arg:

`getter: static (obj) => ((global::System.Collections.DictionaryEntry)obj).Value,`

In this case, `obj.Value` is nullable, but the getter type (`Func<object, T>?`) is expecting a non-null return value; hence the error. 

In this PR I fixed it by using `!`, but also considered changing the getter parameter type to `Func<object, T?>?`. I ran into some issues with that which I try to explain in https://github.com/dotnet/runtime/issues/52227#issuecomment-896376092.

Reply:
Here's an image showing the problem with the initial [prototype](https://github.com/layomia/dotnet_runtime/commit/6bbf1b640889bec36a531912aa90bdd226380c51#) where I tried to refactor the underlying serializer infra to use `T?` instead of `T`. I don't understand why this error occurs (both in the command line and in VS), so I'll try to make a standalone repro and open an issue in the Roslyn repro.

![cscerror](https://user-images.githubusercontent.com/9868994/129096619-aa7df327-9f00-4033-83a7-83dab98d91e6.png)

For this PR, I just pushed a new commit using the correct signature for the getter/setter parameters after @eiriktsarpalis shared a workaround with me.